chore: Add release notes for Metals 1.1.1 (#5929)
scalameta/metals: website/blog/2023-12-08-silver.md

---
author: Tomasz Godzik
title: Metals v1.1.1 - Silver
authorURL: https://twitter.com/<login>
authorImageURL: https://github.com/<login>.png
---

We're happy to announce the release of Metals v1.1.1, which continues to improve

> reporting mechanisms and is finally feature complete when it comes to Scala 3.

Reviewer: ```suggestion
reporting mechanisms and is finally feature complete when it comes to Scala 3.
```

I think it might be worth explaining what is meant by "feature complete when it comes to Scala 3". Surely I don't assume this means that there will be no new features related to Scala 3 right? Metals will always continue to improve and add new things. I'm assuming this instead means that you've reached feature parity with Scala 2 support?

Author: Changed, should make sense now, let me know what you think!

This means that most new releases will focus on stability instead of adding new

> features, unless those features contribute to making Metals more relaible.

Reviewer 2: ```suggestion
features, unless those features contribute to making Metals more reliable.
```

Author: Fixed

## Use presentation compiler for synthetic decorations

Up until this release Metals would use the semanticdb information for displaying
synthetic decorations such as inferred types or implicit conversions. This was
not ideal as it meant that the information was not always up to date and could
be inconsistent with the current state of code.

![synthetics](https://i.imgur.com/td11vRH.gif)

To fix that [jkciesluk](https://github.com/jkciesluk) switched to using the
presentation compiler for this information. This means that each time code
changes the decorations will be refreshed. This also added a number of new
decorations that were not previously available for Scala 3 and it will be on par
with the Scala 2 ones.

## Base Metals view on indexing information

Metals tree view enables users to navigate their workspace in a more structured
way by directly showing the build modules and their packages, classes, objects
etc. It would also allow users to navigate their dependency jars. Previously,
this was based on special Scala 2 signature information, which means it wouldn't
work in Scala 3.

From this release we changed it to use the same information that is used for

> indexing in case of jars and semanticdb for local workspace files. This means

Reviewer: If this uses the same information, does this also improve the performance? If so, it might be nice to add that as well.

Author: It's very similar I think in terms of performance (maybe a bit faster), we do need to index it twice, since the information we need is totally opposite than what we need from normal indexes. And I didn't want to make global indexes much larger. Though here we use the same fast indexing mechanism instead of more costly loading of classfiles previously.

With normal indexes we need `symbol -> file` and here we need `file -> symbol`. Creating that info from existing indexes was slow since we needed to search the whole map of dependencies for symbols in a particular file. If we saved that in memory, we would have a double the memory usage for indexes.

Author: Checking again, the performance might sometimes be a bit worse, but it's only noticeable in larger jars (and still not much slower). I can make it faster but the most important thing here was making sure it always work reliably. Also, I think the biggest issue with performance is not actually connected to the method of indexing.

I will take a look at it at some point.

that Scala 3 is fully supported and this has been the last big feature missing
for full parity between Scala 2 and 3.

![tree-view](https://i.imgur.com/Vun4kJh.gif)

All the existing features should now work for Scala 3 as well. This includes the
reveal command which allows to quickly reveal the current file in the tree view
including inside jars.

![tree-view-reveal](https://i.imgur.com/xW0W3I3.gif) Additionally, we improved
icons used by the tree view. They will now use the default VS Code icons for the

> most part, which should make them more consistent with the rest of the editor.

Reviewer: So I totally missed this change and now realize this is now broken in nvim-metals. Technically this is a breaking change for any client implementing the tree view protocol, since we now change what the `icon` names are. So in my case, no icons are showing since I had a mapping from the old names to the icons that corresponded with them.  I thought we weren't doing patch releases when we break things for clients?

I see down below there is a section for plugin authors and we just link to the Scala file with the new names. It'd be even more helpful to have something that lists the old names to new names. That'd make it much easier. For example

```
object -> symbol-object
...
```

Then authors don't need to dig through [this](https://github.com/scalameta/metals/commit/054fb2aa60650dcff089714efd271fa5c15cf22a#diff-168043e522e5797860de334e85bcb3dbce72414f4fc906ddd4928023d35182d6) to see what changed. I'd also recommend making mention that these icons correspond to the VS Code icons [here in the docs](https://scalameta.org/metals/docs/integrations/tree-view-protocol#tree-view-node).

Author: > Technically this is a breaking change for any client implementing the tree view protocol, since we now change what the icon names are. So in my case, no icons are showing since I had a mapping from the old names to the icons that corresponded with them. I thought we weren't doing patch releases when we break things for clients?

Ach sorry, I didn't check and I assumed that only VS Code used icons.

It will be easiest to revert the changes to icons and translate them in VS Code. We wouldn't need to change anything.

Reviewer: 😢 always forgetting that nvim has nice things 😢

Author: I forgot I also added some more types and I think it all is more consistent than before. Maybe we'll just do 1.2.0, because of that? What do you think?

Author: > 😢 always forgetting that nvim has nice things 😢

I am so sorry, I just thought it's terminal oriented so it's harder to get graphical stuff right. That was also due to my laziness :(

Author: Changed to 1.2.0

## Go to definition for scaladoc

When writing scaladoc for your methods or classes you can use links, which are
of a form `[[name.name.ClassName]]`. Thanks to
[kasiaMarek](https://github.com/kasiaMarek) Metals will now allow you to
navigate to the target of the link via normal go to definition.

![scaladoc-def](https://i.imgur.com/TYNMCEm.gif)

To learn more about linking in scaladoc check out the
[Scala documentation](https://docs.scala-lang.org/scala3/guides/scaladoc/linking.html)

## Custom project root setting

Metals currently will try to find the proper build tool root for your project,
which is useful especially if you open a parent directory that doesn't contain a
build tool configuration. However, this can be problematic if you have a less
traditional setup and you need to be able to control the root directory. For
example if you have an sbt source dependency and you still want to have that
dependency open in the same window.

> To help with that we added a new setting `metals.customProjectRoot` that allows

Reviewer: Not breaking, but maybe mention this down below for plugin authors as well to draw extra attention that they should add it, since I haven't added it yet to nvim-metals even 😬

Author: Done

you to specify the root directory for your project. This setting is available in
the settings UI and can be set to a path relative to the workspace root.

![custom-root](https://i.imgur.com/sgr1j1T.png)

Another amazing feature from [kasiaMarek](https://github.com/kasiaMarek)

## Bsp errors in status

In the previous release Metals started showing more errors from the build server
with the intention on making it clearer to the user what is going on. However,
showing a message request for the user could be quite invasive and not always
relevant.

To help with that [kasiaMarek](https://github.com/kasiaMarek) added a new status
bar item that will show the number of errors in the build server. This will be

> shown only if there are any errors and will be hidden once the reports connected

Reviewer: Is there needed for clients to get this new status item? If `bspStatusBarProvider` is `"on"` and implemented then it should _just work_ ™️ right? Or is there any actual changes to the status that need to be account for?

contributors that require it.

## Changes for plugin authors

Tree view icons changed and use a number of default VS Code ones. If previously
icons were show in other editors, this change could cause them to stop being
shown properly. Full list of icon names can be found
[here](https://github.com/scalameta/metals/blob/main/metals/src/main/scala/scala/meta/internal/tvp/ClasspathTreeView.scala#L109).

## Contributors

Big thanks to everybody who contributed to this release or reported an issue!

```
$ git shortlog -sn --no-merges v1.1.0..v1.1.1
    40	Tomasz Godzik
    28	Katarzyna Marek
    27	Jakub Ciesluk
    22	Scalameta Bot
     7	rochala
     4	scalameta-bot
     3	dependabot[bot]
     2	Chris Kipp
     2	Karan Ahlawat
     2	Lucas Nouguier
     2	tgodzik
     1	Arthur McGibbon
     1	Jamie Thompson
     1	Lorenzo Gabriele
     1	Seth Tisue
     1	naferx
```

## Miscellaneous

This Metals version includes a large number of fixes and improvements and a lot
of them could improve your experience significantly. We highly recommend looking
at the list below to see what changed. Please remember that you can always reach

> out to use and will will try to fix any issues you might have.

Reviewer 2: ```suggestion
out to us and we will try to fix any issues you might have.
```
